feat: google_drive write tools and read comment tool

sheagcraig opened the pull request:
Hi friends!
This PR adds the ability to the Gdrive MCP server to upload files (just whatever you want) and to create and update Google Docs, Sheets, and Slides.

I have some other tools coming, but I'm waiting on perms from our Google admins for the app, so I figured I'd start the process of getting this reviewed.

kalvinnchau commented on crates/goose-mcp/src/google_drive/mod.rs:
```diff
@@ -232,6 +236,198 @@ impl GoogleDriveRouter {
+                      "description": "Path to a local file to use to update the Google Drive file. Mutually exclusive with body.",
+                  }
+              },
+              "required": ["name", "mimeType"],
```
looks like it should be `fileId` that's required, not `name`

kalvinnchau commented on crates/goose-mcp/src/google_drive/mod.rs:
```diff
@@ -810,6 +1018,394 @@ impl GoogleDriveRouter {
+            .await;
+        match result {
+            Err(e) => Err(ToolError::ExecutionError(format!(
+                "Failed to updategoogle doc {}, {}.",
```
```suggestion
                "Failed to update google doc {}, {}.",
```

kalvinnchau commented on crates/goose-mcp/src/google_drive/mod.rs:
```diff
@@ -810,6 +1018,394 @@ impl GoogleDriveRouter {
+            .await;
+        match result {
+            Err(e) => Err(ToolError::ExecutionError(format!(
+                "Failed to updategoogle slides {}, {}.",
```
```suggestion
                "Failed to update google slides {}, {}.",
```

kalvinnchau commented on crates/goose-mcp/src/google_drive/mod.rs:
```diff
@@ -810,6 +1018,394 @@ impl GoogleDriveRouter {
+                .get("fileId")
+                .and_then(|q| q.as_str())
+                .ok_or(ToolError::InvalidParameters(
+                    "The fileIdparam is required".to_string(),
```
```suggestion
                    "The fileId param is required".to_string(),
```

kalvinnchau commented on crates/goose-mcp/src/google_drive/mod.rs:
```diff
@@ -810,6 +1018,394 @@ impl GoogleDriveRouter {
+                .get("fileId")
+                .and_then(|q| q.as_str())
+                .ok_or(ToolError::InvalidParameters(
+                    "The fileIdparam is required".to_string(),
```
```suggestion
                    "The fileId param is required".to_string(),
```

kalvinnchau reviewed (commented):
Awesome work!

Instead of defaulting to true for:
```rust
            .supports_team_drives(true)
            .supports_all_drives(true)
```
Should all the write operations also take in a parameter about the drive its writing to (like the `corpora` in search) so that by  default we only write to a user's personal drive and only write to shared drives when explicitly asked to? 


I noticed a bunch of duplication for upload/create files, could we extract that to some helper function and pass in params?

something like this?
```rust
enum FileOperation {
    Create { name: String },
    Update { file_id: String },
}

async fn upload_file(
    &self,
    operation: FileOperation,
    content: Box<dyn ReadSeek>,
    source_mime_type: &str,
    target_mime_type: &str,
) -> Result<Vec<Content>, ToolError> {
// body
}

// then upload/upload_doc/create_doc:
async fn upload(&self, params: Value) -> Result<Vec<Content>, ToolError> {
    //get params/content

    self.upload_file(
        FileOperation::Create { name: name.to_string() },
        content,
        mime_type,
        mime_type
    ).await

async fn create_doc(&self, params: Value) -> Result<Vec<Content>, ToolError> {
   // get params

    let source_mime_type = "text/markdown";
    let target_mime_type = "application/vnd.google-apps.document";
    let content = Box::new(Cursor::new(body.as_bytes()));

    self.upload_file(
        FileOperation::Create { name: name.to_string() },
        content,
        source_mime_type,
        target_mime_type
    ).await
}
```

sheagcraig commented:
@kalvinnchau thanks for catching all of those dumb mistakes. I've been running Zed instead of vim and I think it's doing some completions for me that are making me whiff on things like that.

I'll incorporate all of these tomorrow morning.

sheagcraig commented:
Okay-I updated to add a bool whether you want to allow the tool to access just your drive, or all of your available drives.

I also added a `parent` arg which lets you specify a folder ID to save a newly created folder in (didn't do this for updates because that works differently, and I'm going to write a move func.

So that's great! It defaults to _not_ using "all drives", and it defaults to saving at the root of your drive.

The only problem is, even just doing this in the MCP inspector, I can't get it to actually upload files into another "parent" folder. I tried folders that are in "My Drive" as well as shared ones to no avail. So probably something we'll have to experiment with down the road.

kalvinnchau reviewed (approved):
🔥
